site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -2339,6 +2339,84 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
         }
     }
 
+    /**
+     * Route for saving the marks the submitter received for a component
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/change_grade_version", methods: ["POST"])]
+    public function ajaxUpdateGradedVersionForStudent(string $gradeable_id): void {
+        $anon_id = $_POST['anon_id'] ?? null;
+        $graded_version = intval($_POST['graded_version'] ?? null);
+        $component_ids = $_POST['component_ids'] ?? [];
+
+        if ($anon_id === null) {
+            $this->core->getOutput()->renderJsonFail('Missing anon_id parameter');
+            return;
```

Comment:
It would be better to return `JsonResponse::getFailResponse(...)` instead of calling renderJsonFail directly.  Apply throughout.